velox/connectors/hive/storage_adapters/s3fs/RegisterS3FileSystem.cpp

```diff
@@ -26,73 +26,89 @@
-                                       std::string_view filePath) {
-    folly::call_once(S3FSInstantiationFlag, [&properties]() {
-      std::shared_ptr<S3FileSystem> fs;
-      if (properties != nullptr) {
-        initializeS3(properties.get());
-        fs = std::make_shared<S3FileSystem>(properties);
-      } else {
-        auto config = std::make_shared<core::MemConfig>();
-        initializeS3(config.get());
-        fs = std::make_shared<S3FileSystem>(config);
-      }
-      s3fs = fs;
-    });
-    return s3fs;
-  };
-  return filesystemGenerator;
+std::string getS3Identity(std::shared_ptr<core::MemConfig>& config) {
+  HiveConfig hiveConfig = HiveConfig(config);
+  auto endpoint = hiveConfig.s3Endpoint();
+  if (!endpoint.empty()) {
+    // The identity is the endpoint.
+    return endpoint;
+  }
+  // Default key value.
+  return "aws-s3-key";
 }
 
-std::function<std::unique_ptr<velox::dwio::common::FileSink>(
-    const std::string&,
-    const velox::dwio::common::FileSink::Options& options)>
-s3WriteFileSinkGenerator() {
-  static auto s3WriteFileSink =
-      [](const std::string& fileURI,
-         const velox::dwio::common::FileSink::Options& options)
-      -> std::unique_ptr<dwio::common::WriteFileSink> {
-    if (isS3File(fileURI)) {
-      auto fileSystem =
-          filesystems::getFileSystem(fileURI, options.connectorProperties);
-      return std::make_unique<dwio::common::WriteFileSink>(
-          fileSystem->openFileForWrite(fileURI, {{}, options.pool}),
-          fileURI,
-          options.metricLogger,
-          options.stats);
-    }
-    return nullptr;
-  };
+std::shared_ptr<FileSystem> fileSystemGenerator(
+    std::shared_ptr<const Config> properties,
+    std::string_view /*filePath*/) {
+  std::shared_ptr<core::MemConfig> config = std::make_shared<core::MemConfig>();
+  if (properties) {
+    *config = core::MemConfig(properties->values());
+  }
+  const auto s3Identity = getS3Identity(config);
+
+  return fileSystems().withWLock(
+      [&](auto& instanceMap) -> std::shared_ptr<FileSystem> {
+        initializeS3(config.get());
+        if (instanceMap.count(s3Identity) == 0) {
+          auto fs = std::make_shared<S3FileSystem>(properties);
+          instanceMap.insert({s3Identity, fs});
+          return fs;
+        }
+        return instanceMap[s3Identity];
+      });
+}
 
-  return s3WriteFileSink;
+std::unique_ptr<velox::dwio::common::FileSink> s3WriteFileSinkGenerator(
+    const std::string& fileURI,
+    const velox::dwio::common::FileSink::Options& options) {
+  if (isS3File(fileURI)) {
+    auto fileSystem =
+        filesystems::getFileSystem(fileURI, options.connectorProperties);
+    return std::make_unique<dwio::common::WriteFileSink>(
+        fileSystem->openFileForWrite(fileURI, {{}, options.pool}),
+        fileURI,
+        options.metricLogger,
+        options.stats);
+  }
+  return nullptr;
 }
 #endif
 
 void registerS3FileSystem() {
 #ifdef VELOX_ENABLE_S3
```

Comment:
probably subject for another PR, but shouldn't these things throw some sort of error if clients try to registerS3FileSystems but S3 support is not compiled in?

Reply:
We should. I will add in a separate PR.